src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionTextCache.cs

```diff
@@ -0,0 +1,108 @@
+
+        private class LambdaExpressionComparer : IEqualityComparer<LambdaExpression>
+        {
+            public static readonly LambdaExpressionComparer Instance = new LambdaExpressionComparer();
+
+            public bool Equals(LambdaExpression lambdaExpression1, LambdaExpression lambdaExpression2)
+            {
+                // We will cache only pure member access expressions. Hence we compare two expressions 
+                // to be equal only if they are identical member access expressions.
+                var expression1 = lambdaExpression1.Body;
+                var expression2 = lambdaExpression2.Body;
+                var identical = false;
+
+                while (true)
+                {
+                    if (expression1 == null || expression2 == null)
+                    {
+                        break;
+                    }
+
+                    if (expression1.NodeType != expression2.NodeType)
+                    {
+                        break;
+                    }
+
+                    if (expression1.NodeType == ExpressionType.MemberAccess)
+                    {
+                        var memberExpression1 = (MemberExpression)expression1;
+                        var memberName1 = memberExpression1.Member.Name;
+                        expression1 = memberExpression1.Expression;
+
+                        var memberExpression2 = (MemberExpression)expression2;
+                        var memberName2 = memberExpression2.Member.Name;
+                        expression2 = memberExpression2.Expression;
+
+                        if (memberName1.Contains("__") && memberName2.Contains("__"))
+                        {
+                            identical = true;
+                            break;
+                        }
+
+                        if (!memberName1.Equals(memberName2))
+                        {
+                            break;
+                        }
+                    }
+                    else 
+                    {
+                        identical = true;
+                        break;
+                    }
+                }
+
+                return identical;
+            }
+
+            public int GetHashCode(LambdaExpression lambdaExpression)
+            {
+                var expression = lambdaExpression.Body;
+                var hashCodeCombiner = HashCodeCombiner.Start();
+
+                while (true)
+                {
+                    if (expression != null && expression.NodeType == ExpressionType.MemberAccess)
+                    {
+                        var memberExpression = (MemberExpression)expression;
+                        var memberName = memberExpression.Member.Name;
+                        // If identifier contains "__", it is "reserved for use by the implementation" and likely compiler-
+                        // or Razor-generated e.g. the name of a field in a delegate's generated class.
```

Comment:
Move these comments above [line 57](https://github.com/aspnet/Mvc/pull/4298/files#diff-cbc498e6135dcff3a95d99f007858dd4R57). Good to have just one comment like this but it should be w/ the first related code.